clash_lib/src/proxy/tuic/mod.rs

```diff
@@ -0,0 +1,296 @@
+        )?;
+        endpoint.set_default_client_config(quinn_config);
+        let endpoint = TuicEndpoint {
+            ep: endpoint,
+            server: ServerAddr::new(opts.server.clone(), opts.port, None),
+            uuid: opts.uuid,
+            password: Arc::from(opts.password.clone().into_bytes().into_boxed_slice()),
+            udp_relay_mode: types::UdpRelayMode::Native,
+            zero_rtt_handshake: opts.reduct_rtt,
+            heartbeat: opts.heartbeat_interval,
+            gc_interval: opts.gc_interval,
+            gc_lifetime: opts.gc_lifetime,
+        };
+        Ok(Arc::new(Self {
+            opts,
+            ep: endpoint,
+            conn: AsyncMutex::new(None),
+            next_assoc_id: AtomicU16::new(0)
+        }))
+    }
+    async fn get_conn(&self) -> Result<TuicConnection> {
+        let fut = async {
+            let mut guard = self.conn.lock().await;
+            if guard.is_none() {
+                // init
+                *guard = Some(self.ep.connect().await?);
+            }
+            let conn = guard.take().unwrap();
+            let conn = if conn.is_closed() {
+                // reconnect
+                self.ep.connect().await?
+            } else {
+                conn
+            };
+            // TODO TuicConnection is huge, is it necessary to clone it? If it is, should we use Arc ?
+            *guard = Some(conn.clone());
+            Ok(conn)
+        };
+        tokio::time::timeout(self.opts.request_timeout, fut).await?
+    }
+
+    async fn do_connect_stream(
+        &self,
+        sess: &Session,
+        _resolver: ThreadSafeDNSResolver,
+    ) -> Result<BoxedChainedStream> {
+        let conn = self.get_conn().await?;
+        let dest = sess.destination.clone().into_tuic();
+        let tuic_tcp = conn.connect_tcp(dest).await?.compat();
+
+        let s = ChainedStreamWrapper::new(tuic_tcp);
+        s.append_to_chain(self.name()).await;
+        Ok(Box::new(s))
+    }
+
+    async fn do_connect_datagram(
+        &self,
+        sess: &Session,
+        _resolver: ThreadSafeDNSResolver,
+    ) -> Result<BoxedChainedDatagram> {
+        let conn = self.get_conn().await?;
+
+        let assos_id = self.next_assoc_id.fetch_add(1, Ordering::Relaxed);
+        let quic_udp = TuicDatagramOutbound::new(assos_id, conn, sess.source.into());
+        let s = ChainedDatagramWrapper::new(quic_udp);
+        s.append_to_chain(self.name()).await;
+        Ok(Box::new(s))
+    }
+}
+
+struct TuicDatagramOutbound {
+    assoc_id: u16,
+    // conn: TuicConnection,
+    // dest: tuic::Address,
+    handle: tokio::task::JoinHandle<Result<()>>,
+    send_tx: tokio_util::sync::PollSender<UdpPacket>,
+    recv_rx: tokio::sync::mpsc::Receiver<UdpPacket>,
+}
+
+impl TuicDatagramOutbound {
+    pub fn new(
+        assoc_id: u16,
+        conn: TuicConnection,
+        local_addr: ClashSocksAddr,
+    ) -> AnyOutboundDatagram {
+        // TODO not sure about the size of buffer
```

Comment:
we should probably implement buff pool to avoid alloc.

there are a few places like this in the code base, but let's keep this for now.

Reply:
True, backpressure is a pretty tricky thing